connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,15 +290,54 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
             throw new IllegalStateException("At least one non-null offset store must be provided");
         }
 
-        return primaryStore.set(values, (primaryWriteError, ignored) -> {
+        List<ByteBuffer> partitionsWithTombstoneOffsets = values.entrySet().stream()
+                .filter(offsetEntry -> offsetEntry.getValue() == null)
+                .map(Map.Entry::getKey).collect(Collectors.toList());
+
+        Map<ByteBuffer, ByteBuffer> tombstoneOffsets = new HashMap<>();
+        for (ByteBuffer partition : partitionsWithTombstoneOffsets) {
+            tombstoneOffsets.put(partition, null);
+        }
+        Map<ByteBuffer, ByteBuffer> regularOffsets = values.entrySet().stream()
+                .filter(offsetEntry -> offsetEntry.getValue() != null)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        // If the supplied offsets contain tombstone values, then tombstone offsets are extracted first,
+        // written to the secondary store in a synchronous manner and then to the primary store.
+        // This is because, if a tombstone offset is successfully written to the per-connector offsets topic,
+        // but cannot be written to the global offsets topic, then the global offsets topic will still contain that
+        // source offset, but the per-connector topic will not. Due to the fallback-on-global logic used by the worker,
+        // if a task requests offsets for one of the tombstoned partitions, the worker will provide it with the
+        // offsets present in the global offsets topic, instead of indicating to the task that no offsets can be found.
+        CompletableFuture<Void> offsetWriteFuture = CompletableFuture.completedFuture(null);
+        if (secondaryStore != null && !tombstoneOffsets.isEmpty()) {
+            offsetWriteFuture.thenAccept(v -> {
```

Comment:
This causes the passed-in lambda to be executed after `offsetWriteFuture` completes, but it doesn't cause `offsetWriteFuture::get` to block on the completion of that lambda, which means that if the caller of `ConnectorOffsetBackingStore::set` waits for the returned `Future` to complete right now, it'll complete immediately even though it's likely no offsets will have been written to either the primary or secondary store.

Also, nit: when researching this behavior, I ran into `CompletableFuture::thenRun`, which is a slightly better fit than `CompletableFuture::thenAccept` (since we don't use the return value).

```suggestion
            offsetWriteFuture = offsetWriteFuture.thenRun(() -> {
```

Reply:
I have actually removed the chaining using `CompletableFuture` and simplified the logic. I just wait on the secondary store write directly (with or without a timeout) and if the execution fails or the wait itself fails, I update the callback and return the same exception (because it's already completed).

```
if (secondaryStore != null && !tombstoneOffsets.isEmpty()) {
            Future<Void> secondaryWriteFuture = secondaryStore.set(tombstoneOffsets, (t, r) -> { });
            try {
                if (exactlyOnce) {
                    secondaryWriteFuture.get();
                } else {
                    secondaryWriteFuture.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
                }
                log.debug("Successfully flushed tombstone offsets to secondary store");
            } catch (ExecutionException e) {
                log.error("{} Failed to flush tombstone offsets to secondary store", this, e.getCause());
                callback.onCompletion(e.getCause(), null);
                return secondaryWriteFuture;
            } catch (Throwable e) {
                log.error("{} Failed to flush tombstone offsets to secondary store", this, e);
                callback.onCompletion(e, null);
                return secondaryWriteFuture;
            }
        }
```

With this,  there are no changes (apart from using `regularOffsets` when writing to secondary store) from the primary store. Let me know what you think about this.

Reply:
I re-read the comments again and looks like with this we are going against the approach you wanted in this [comment](https://github.com/apache/kafka/pull/13801/#discussion_r1268520271) i.e 

> That said, I don't love how we've made this method synchronously await the write to the secondary store. We should return a Future to the caller that corresponds to all of the offset flushes that we'd need to block on for an offset commit (i.e., the existing flush that we're performing, possibly preceded by a preemptive flush of tombstones to the secondary store).

Let me take a look at this again. Sorry about this but I had forgotten about this comment in particular.